doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
+
+class Scope final {
+public:
+  Scope();
+  Scope(const std::shared_ptr<Scope>& parent);
+
+  //! Get Variable in this scope.
+  //! @return nullptr if no such variable.
+  const VariablePtr& getVar(const std::string& name) const;
+
+  //! Create or get a variable in this scope.
+  VariablePtr& createOrGetVar(const std::string& name);
+
+private:
+  /// variable name -> variable
+  std::unordered_map<std::string, VariablePtr> vars_;
```

Comment:
Why `unordered_map` other than `map`?  I am not sure which is better -- I'd always believed the former is a hash map and the latter is an RD-tree.

Anyway, I noticed that Caffe2 is using `map` in its [Workspace](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/workspace.h#L216):

```cpp
 private:
  BlobMap blob_map_;
```

where `BlobMap` was defined as a `CaffeMap` at [here](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/workspace.h#L56):

```cpp
typedef CaffeMap<string, unique_ptr<Blob> > BlobMap;
```

and `CaffeMap` is defined [here](https://github.com/caffe2/caffe2/blob/v0.7.0/caffe2/core/common.h#L50) 

```cpp
using CaffeMap = std::map<Key, Value>;
// using CaffeMap = std::unordered_map;
```